File: lib/cfr/builtins/libs/dart/async/completer.dart
import 'dart:async';
import 'dart:core';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedCompleter extends VMManagedBox<Completer<dynamic>> {
  VMManagedCompleter(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['getFuture'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Future>(
            object: vmObject.future,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['complete'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.complete(maybeUnBoxAndBuildArgument<FutureOr<dynamic>?, dynamic>(
          luaCallerArguments[1],
          parentState: hydroState));
      return [];
    });
    table['completeError'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.completeError(
          maybeUnBoxAndBuildArgument<Object, dynamic>(luaCallerArguments[1],
              parentState: hydroState),
          maybeUnBoxAndBuildArgument<StackTrace?, dynamic>(
              luaCallerArguments[2],
              parentState: hydroState));
      return [];
    });
    table['getIsCompleted'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.isCompleted,
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final Completer vmObject;
}

void loadCompleter(
    {required HydroState hydroState, required HydroTable table}) {
  table['completer'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Completer>(
          object: Completer(),
          hydroState: hydroState,
          table: luaCallerArguments[0])
    ];
  });
  table['completerSync'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Completer>(
          object: Completer.sync(),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<Completer>(boxer: (
      {required Completer vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedCompleter(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/async/completer.ts
import { IStackTrace } from "../core/stackTrace";
import { IFuture } from "./future";
import { IFutureOr } from "./futureOr";
declare const dart: {
    async: {
        completer: <T>(this: void, completer: ICompleter<T>) => ICompleter<T>;
        completerSync: <T>() => ICompleter<T>;
    };
};
export interface ICompleter<T> {
    getFuture: () => IFuture<T>;
    complete: (value?: IFutureOr<T> | undefined) => void;
    completeError: (
        error: Object,
        stackTrace?: IStackTrace | undefined
    ) => void;
    getIsCompleted: () => boolean;
}
export class Completer<T> {
    public constructor() {
        dart.async.completer(this);
    }
    public static sync<T>(): ICompleter<T> {
        return dart.async.completerSync();
    }
    private readonly _dart_getFuture: () => IFuture<T> = undefined as any;
    private readonly _dart_complete: (
        value?: IFutureOr<T> | undefined
    ) => void = undefined as any;
    private readonly _dart_completeError: (
        error: Object,
        stackTrace?: IStackTrace | undefined
    ) => void = undefined as any;
    private readonly _dart_getIsCompleted: () => boolean = undefined as any;
    public getFuture(): IFuture<T> {
        return this._dart_getFuture();
    }
    public complete(value?: IFutureOr<T> | undefined): void {
        return this._dart_complete(value);
    }
    public completeError(
        error: Object,
        stackTrace?: IStackTrace | undefined
    ): void {
        return this._dart_completeError(error, stackTrace);
    }
    public getIsCompleted(): boolean {
        return this._dart_getIsCompleted();
    }
}
